sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs

```diff
@@ -659,604 +600,482 @@ protected EventProcessorClient()
+        }
 
-            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
-            IEnumerable<EventProcessorCheckpoint> availableCheckpoints;
+        /// <summary>
+        ///   Performs the tasks to initialize a partition, and its associated context, for event processing.
+        /// </summary>
+        ///
+        /// <param name="partition">The context of the partition being initialized.  Only the well-known members of the <see cref="EventProcessorPartition" /> will be populated.  If a custom context is being used, the implementor of this method is responsible for initializing custom members.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the initialization.  This is most likely to occur if the partition is claimed by another event processor instance or the processor is shutting down.</param>
+        ///
+        /// <remarks>
+        ///   It is not recommended to that the state of the processor be managed directly from within this method; requesting to start or stop the processor may result in
+        ///   a deadlock scenario, especially if using the synchronous form of the call.
+        /// </remarks>
+        ///
+        protected override async Task OnInitializingPartitionAsync(EventProcessorPartition partition,
+                                                                   CancellationToken cancellationToken)
+        {
+            // Handlers cannot be changed while the processor is running; it is safe to check and call
+            // without capturing a local reference.
 
-            try
-            {
-                availableCheckpoints = await StorageManager.ListCheckpointsAsync(FullyQualifiedNamespace, EventHubName, ConsumerGroup, cancellationToken).ConfigureAwait(false);
-            }
-            catch (Exception ex)
+            if (_partitionInitializingAsync != null)
             {
-                cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
-
-                // If processing task creation fails, we'll try again on the next time this method is called.
-                // This should happen on the next load balancing loop as long as this instance still owns the
-                // partition.
+                var eventArgs = new PartitionInitializingEventArgs(partition.PartitionId, DefaultStartingPosition, cancellationToken);
+                await _partitionInitializingAsync(eventArgs).ConfigureAwait(false);
 
-                Logger.StartPartitionProcessingError(partitionId, ex.Message);
-                var errorEventArgs = new ProcessErrorEventArgs(partitionId, Resources.OperationListCheckpoints, ex, cancellationToken);
-                _ = OnProcessErrorAsync(errorEventArgs);
-
-                return;
+                PartitionStartingPositionDefaults[partition.PartitionId] = eventArgs.DefaultStartingPosition;
```

Comment:
Thanks!   It wouldn't have been possible under the original design, if I recall.   We can now cheat since the storage operations are not invoked directly on the storage manager but work through the processor.   I don't believe that we thought of that during the first attempt.